webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ArchiveApi.scala

```diff
@@ -1086,69 +1131,50 @@ class SaveArchiveServicebyRepo(
     )
 
     for {
-      _        <- ApplicationLoggerPure.Archive.debug(
-                    s"Adding technique from archive: '${t.technique.name}' (${techniqueDir.pathAsString})"
-                  )
-      // path are relative to technique. Files we need to add as new in the end
-      addedRef <- Ref.make(Chunk.fromIterable(t.files.map(_._1)))
-      diffRef  <- Ref.make(Chunk[ResourceFile]())
-      // get all file path (relative to technique dir)
-      existing <- IOResult.attempt {
-                    if (techniqueDir.exists) {
-                      techniqueDir
-                        .collectChildren(_ => true)
-                        .toList
-                        .map(_.pathAsString.replaceFirst(techniqueDir.pathAsString + "/", ""))
-                    } else { // technique or technique version does not exists
-                      Chunk.empty
-                    }
-                  }
-      _        <- ZIO.foreachDiscard(existing) { e =>
-                    for {
-                      keep <- addedRef.modify(a => {
-                                if (a.contains(e)) { (true, a.filterNot(_ == e)) }
-                                else { (false, a) }
-                              })
-                      _    <- diffRef.update(_ :+ ResourceFile(e, if (keep) ResourceFileState.Modified else ResourceFileState.Deleted))
-                    } yield ()
-                  }
-      added    <- addedRef.get.map(_.map(ResourceFile(_, ResourceFileState.New)))
-      updated  <- diffRef.get
-
+      _    <- ApplicationLoggerPure.Archive.debug(
+                s"Adding technique from archive: '${t.technique.name}' (${techniqueDir.pathAsString})"
+              )
+      diff <- SaveArchiveServicebyRepo.buildDiff(t, techniqueDir)
       // now, actually delete files marked so
-      _ <- ApplicationLoggerPure.Archive.trace(
-             s"Deleting technique files for technique '${t.technique.id.serialize}': ${updated.collect {
-                 case f if (f.state == ResourceFileState.Deleted) => f.path
-               }.mkString(", ")}"
-           )
-      _ <- ZIO.foreachDiscard(updated) { u =>
-             if (u.state == ResourceFileState.Deleted) {
-               IOResult.attempt(File(techniqueDir.pathAsString).delete())
-             } else ZIO.unit
-           }
+      _    <- ApplicationLoggerPure.Archive.trace {
+                val deleted = diff.collect {
+                  case f if (f.state == ResourceFileState.Deleted) => f.path
+                }.toList match {
+                  case Nil => "none"
+                  case l   => l.mkString(", ")
+                }
+                s"Deleting technique files for technique '${t.technique.id.serialize}': ${deleted}"
+              }
+      _    <- ZIO.foreachDiscard(diff) { u =>
+                if (u.state == ResourceFileState.Deleted) {
+                  IOResult.attempt(File(techniqueDir.pathAsString).delete()).catchSome {
```

Comment:
Have I missed something there because the whole technique directory is deleted ? Shouldn't it be the delete file `u` ?

Reply:
This may also be the reason why the `NoSuchFileException` was raised when more that 1 file was deleted from the archive

Reply:
yes, right!